### DISCUSSION TITLE
Spell Check on non English language

### DISCUSSION THREAD

**Discussion by alvesman** [R1j1t/contextualSpellCheck#54] [Q&A] [answered]
Hi,
I'm not sure if this is a bug or just my lack of skills on using this...
I've successfully run the English spellCheck example. When I use a different language (in this case Portuguese) it seems that the spell check still runs as if the language was English. I've attached a colab notebook showing the “problem”:

```
!pip install spacy
!pip install contextualSpellCheck
!python -m spacy download pt_core_news_sm
import spacy
import contextualSpellCheck

nlp = spacy.load("pt_core_news_sm")
contextualSpellCheck.add_to_pipe(nlp)
print(nlp.pipe_names)
print(nlp.lang)

# There is only one spelling error on the pt text: 'garafa' should be 'garrafa'
texts = [
    "O rato roeu a rolha da garafa.",
    "Net incme was $9.4 million compared to the prior year of $2.7 milion."
]

for text in texts:
    doc = nlp(text)
    print(doc._.performed_spellCheck)
    print(doc._.outcome_spellCheck)
    print(doc._.suggestions_spellCheck)
```
OUTPUT of the corrections
```True
O te que aa dana.
{rato: 'te', roeu: 'que', rolha: '##a', garafa: '##na'}

True
Net income was $9.4 million compared to the prior year of $2.7 million.
{incme: 'income', milion: 'million'}
```
Instead of "O te que aa dana." the output for the first text should be "O rato roeu a rolha da garrafa.". Nevertheless, the English text is corrected. 

Thanks for any clue of what am I doing wrong,
Z.
[SpellCheck_Spacy.zip](https://github.com/R1j1t/contextualSpellCheck/files/6183652/SpellCheck_Spacy.zip)

**Comment by R1j1t** [comment, marked as answer]
Hi, I think I understand the issue. You have passed the language specific spacy model (in this case Portuguese) but for the spell check you are still using the default model (english bert-base-cased). You can try to search for a Portuguese model from HuggingFace Models and pass it ex. https://huggingface.co/neuralmind/bert-base-portuguese-cased. 

Please see an examples for Japanese
https://github.com/R1j1t/contextualSpellCheck/blob/5b65bad18fd848782802ecf0979cb54579b30aec/examples/ja_example.py#L4-L12

I will still verify this in a day or two. If the problem is not resolved then we will have to check for possible solutions!

**Reply by alvesman** [reply]
Hi,
Exactly as you pointed out. Nevertheless I'll try to find another pt trained model. This one does not perform well at spell checking.

Thank you very much fot your time and for sharing!
Best Regards,
Manuel

**Reply by R1j1t** [reply]
I tried using the Portuguese model as below:
```bash
>>> import spacy
>>> import contextualSpellCheck
>>> nlp = spacy.load("pt_core_news_sm")
>>> nlp.add_pipe( 
...      "contextual spellchecker", 
...      config={ 
...          "model_name": "neuralmind/bert-base-portuguese-cased", 
...          "max_edit_dist": 6, 
...      }, 
...  )
>>> doc = nlp("O rato roeu a rolha da garafa.")
>>> doc
O rato roeu a rolha da garafa.
>>> doc.lang_
'pt'
>>> doc._.performed_spellCheck
True
>>> doc._.outcome_spellCheck # expected: O rato roeu a rolha da garrafa.
'O gato tirou a porta da casa.'
```

Outcome is not at all close to the expected. The problems which I think are:

1. unable to identify misspell (Will need to check model's vocab.txt)

 In the above sentence misspell identified were:
```bash
>>> doc._.suggestions_spellCheck
{rato: 'gato', roeu: 'tirou', rolha: 'porta', garafa: 'casa'}
```
2. Suggestions were incorrect. (Depends on model's ability to fill `[MASK]`)
```bash
>>> doc._.score_spellCheck
{rato: [('homem', 0.04462), ('gato', 0.03421), ('vento', 0.03365), ('carro', 0.02961), ('cara', 0.02905), ('rapaz', 0.0233), ('cachorro', 0.0206), ('sujeito', 0.01899), ('animal', 0.01692), ('cão', 0.01618)], roeu: [('é', 0.0888), ('tirou', 0.05159), ('segura', 0.04494), ('abriu', 0.03968), ('tira', 0.03717), ('acertou', 0.02535), ('foi', 0.02388), ('pegou', 0.02378), ('abre', 0.02329), ('quebrou', 0.02203)], rolha: [('porta', 0.24605), ('parede', 0.07212), ('janela', 0.04938), ('cabeça', 0.04437), ('borda', 0.04216), ('grama', 0.03077), ('madeira', 0.0217), ('poeira', 0.02135), ('pedra', 0.01607), ('cerca', 0.01421)], garafa: [('casa', 0.04409), ('árvore', 0.03589), ('porta', 0.03583), ('janela', 0.02902), ('bola', 0.02541), ('cama', 0.02506), ('mesa', 0.0246), ('rede', 0.01802), ('cabeça', 0.01799), ('madeira', 0.01696)]}
```

I am not familiar with the language but if I get time I will try to investigate this.

**Comment by Economax** [comment]
Any updates here?

I need it for Norwegian, if that's any help :)

**Comment by RaulKite** [comment]
Trying to  do spellchecking in Spanish and its not working neither :(

```
nlp = spacy.load("es_dep_news_trf")
nlp.add_pipe(
        "contextual spellchecker",
        config={
          "model_name": "dccuchile/bert-base-spanish-wwm-cased",
          "max_edit_dist": 2,
        },
      )


texto = "¿CUÁNTOS DÍAS ANTICIPA QUE ESTARÁ CERRADO?"

doc=nlp(texto)
print(doc._.performed_spellCheck) # False
print(doc._.outcome_spellCheck) # 
```

Any tip about this?
